File: src/main/modules/leveling.ts
// leveling module — level-ups + AA gains + AA spends + AA-potion quaffs in log order. All
// four are append-only; the delta carries whatever was appended to each since the last
// flush and the renderer concats each array. (The unspent/net-spent math the view
// does depends on log order, which append-only concat preserves.)
//
// THE POTION IS NOT PERSISTED and deliberately so: the quaff is a LOG LINE, so a relaunch
// replays it and re-derives the charge state exactly. A copy in the store would be a second
// source of truth that could disagree with the log after a rescan or an epoch.

import type { EqModule } from './types'
import type { LogEvent } from '../../shared/logEvents'
import type {
  AAEvent,
  AAPotionEvent,
  AASpendEvent,
  LevelEvent,
  LevelingDelta,
  LevelingSnap
} from '../../shared/types'

export class LevelingModule implements EqModule<LevelingSnap, LevelingDelta> {
  readonly id = 'leveling'
  private levels: LevelEvent[] = []
  private aaGains: AAEvent[] = []
  private aaSpends: AASpendEvent[] = []
  private aaPotions: AAPotionEvent[] = []
  private seq = 0
  private pLevels: LevelEvent[] = []
  private pGains: AAEvent[] = []
  private pSpends: AASpendEvent[] = []
  private pPotions: AAPotionEvent[] = []

  reset(): void {
    this.levels = []
    this.aaGains = []
    this.aaSpends = []
    this.aaPotions = []
    this.seq = 0
    this.pLevels = []
    this.pGains = []
    this.pSpends = []
    this.pPotions = []
  }

  onEvent(ev: LogEvent): void {
    this.seq = ev.seq
    switch (ev.kind) {
      case 'epoch':
        // Character rebirth (Task #49): the prior epoch's levels/AA belong to a dead
        // same-name character. Drop them so the AA identity (allocated/unspent/earned) and
        // the level timeline reflect ONLY the current character. Pending deltas are cleared
        // too — during a rescan this happens mid-replay (no deltas push); on a live wipe the
        // renderer re-hydrates from the post-epoch snapshot (index.ts re-sends onCharacter).
        this.levels = []
        this.aaGains = []
        this.aaSpends = []
        this.aaPotions = []
        this.pLevels = []
        this.pGains = []
        this.pSpends = []
        this.pPotions = []
        return
      case 'level': {
        const e: LevelEvent = { ts: ev.ts, level: ev.level }
        this.levels.push(e)
        this.pLevels.push(e)
        return
      }
      case 'aaGain': {
        const e: AAEvent = { ts: ev.ts, amount: ev.amount, nowHave: ev.nowHave }
        this.aaGains.push(e)
        this.pGains.push(e)
        return
      }
      case 'aaSpend': {
        const e: AASpendEvent = { ts: ev.ts, ability: ev.ability, cost: ev.cost, rank: ev.rank }
        this.aaSpends.push(e)
        this.pSpends.push(e)
        return
      }
      case 'aaPotion': {
        const e: AAPotionEvent = { ts: ev.ts }
        this.aaPotions.push(e)
        this.pPotions.push(e)
        return
      }
      default:
        return
    }
  }

  snapshot(): { seq: number; state: LevelingSnap } {
    return {
      seq: this.seq,
      state: {
        levels: this.levels,
        aaGains: this.aaGains,
        aaSpends: this.aaSpends,
        aaPotions: this.aaPotions
      }
    }
  }

  flushDelta(): { seq: number; delta: LevelingDelta } | null {
    if (
      this.pLevels.length === 0 &&
      this.pGains.length === 0 &&
      this.pSpends.length === 0 &&
      this.pPotions.length === 0
    ) {
      return null
    }
    const delta: LevelingDelta = {
      levels: this.pLevels,
      aaGains: this.pGains,
      aaSpends: this.pSpends,
      aaPotions: this.pPotions
    }
    this.pLevels = []
    this.pGains = []
    this.pSpends = []
    this.pPotions = []
    return { seq: this.seq, delta }
  }
}

File: engine/crates/fold/src/modules/leveling.rs
//! `src/main/modules/leveling.ts` — level-ups, AA gains, AA spends and AA-potion quaffs, all four
//! append-only and in log order (the unspent/net-spent math the view does depends on that order).
//!
//! THE POTION IS NOT PERSISTED, deliberately: the quaff is a LOG LINE, so a relaunch replays it
//! and re-derives the charge state exactly.

use crate::event::Event;
use crate::EqModule;
use serde::Serialize;
use serde_json::{json, Value};

#[derive(Debug, Clone, Serialize)]
pub struct LevelRow {
    ts: i64,
    level: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AaGainRow {
    ts: i64,
    amount: i64,
    now_have: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct AaSpendRow {
    ts: i64,
    ability: String,
    cost: i64,
    /// Only the `You have improved X <rank>` shape states one; omitted otherwise.
    #[serde(skip_serializing_if = "Option::is_none")]
    rank: Option<i64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct AaPotionRow {
    ts: i64,
}

#[derive(Default)]
pub struct LevelingModule {
    levels: Vec<LevelRow>,
    aa_gains: Vec<AaGainRow>,
    aa_spends: Vec<AaSpendRow>,
    aa_potions: Vec<AaPotionRow>,
    seq: i64,
    /// THE ANNOUNCE CURSOR (JOS-509) — see [`crate::announce`]. FOUR ARMS APPEND AND THE FIFTH
    /// CLEARS, and that is the whole of what this module publishes; every other line in the log —
    /// which on a levelling grind is essentially all of them — leaves it untouched. This module is
    /// the ticket's named symptom: the leveling scroll hitch was this panel re-fetching its whole
    /// history ten times a second because a mob was being hit.
    announce: crate::announce::Announce,
}

impl LevelingModule {
    pub fn new() -> Self {
        Self::default()
    }
}

impl EqModule for LevelingModule {
    fn id(&self) -> &'static str {
        "leveling"
    }

    fn reset(&mut self) {
        self.levels.clear();
        self.aa_gains.clear();
        self.aa_spends.clear();
        self.aa_potions.clear();
        self.seq = 0;
        self.announce.reset();
    }

    fn on_event(&mut self, ev: &Event, _live: bool) {
        self.seq = ev.seq();
        // ONE BUMP FOR FIVE ARMS, and it is placed after the match rather than inside each of them
        // because the arms are exactly the mutations: four appends and a clear. `_ => {}` is every
        // other line in the log and reaches nothing.
        let published = matches!(
            ev.kind(),
            "epoch" | "level" | "aaGain" | "aaSpend" | "aaPotion"
        );
        if published {
            self.announce.changed(self.seq);
        }
        match ev.kind() {
            // Character rebirth (Task #49): the prior epoch's levels/AA belong to a dead same-name
            // character, and the AA identity (allocated/unspent/earned) is only true without them.
            "epoch" => {
                self.levels.clear();
                self.aa_gains.clear();
                self.aa_spends.clear();
                self.aa_potions.clear();
            }
            "level" => self.levels.push(LevelRow {
                ts: ev.ts(),
                level: ev.int("level").unwrap_or(0),
            }),
            "aaGain" => self.aa_gains.push(AaGainRow {
                ts: ev.ts(),
                amount: ev.int("amount").unwrap_or(0),
                now_have: ev.int("nowHave").unwrap_or(0),
            }),
            "aaSpend" => self.aa_spends.push(AaSpendRow {
                ts: ev.ts(),
                ability: ev.str("ability").unwrap_or_default().to_string(),
                cost: ev.int("cost").unwrap_or(0),
                rank: ev.int("rank"),
            }),
            "aaPotion" => self.aa_potions.push(AaPotionRow { ts: ev.ts() }),
            _ => {}
        }
    }

    /// THE DIRTY BIT (JOS-487, made honest by JOS-509) — a ding, an AA line or a rebirth. See the
    /// `announce` field and `crate::announce`.
    fn published_seq(&self) -> Option<i64> {
        Some(self.announce.cursor())
    }

    fn snapshot(&self) -> Value {
        json!({
            "seq": self.seq,
            "state": {
                "levels": self.levels,
                "aaGains": self.aa_gains,
                "aaSpends": self.aa_spends,
                "aaPotions": self.aa_potions
            }
        })
    }
}
